crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -144,10 +139,25 @@ where
     );
     let piece_getter = Arc::new(FarmerPieceGetter::new(
         NodePieceGetter::new(piece_provider),
-        piece_cache,
-        node.clone(),
+        piece_cache.clone(),
     ));
 
+    let (run_cache_sync_tx, run_cache_sync_rx) = mpsc::channel::<SegmentIndex>(1);
+
+    tokio::spawn({
+        let piece_cache = piece_cache.clone();
+        let piece_getter = piece_getter.clone();
+
+        restore_pieces_cache(run_cache_sync_rx, piece_getter, piece_cache)
+    });
+
+    tokio::spawn({
+        let piece_cache = piece_cache.clone();
+        let node_client = node_client.clone();
+
+        fill_piece_cache_from_archived_segments(run_cache_sync_tx, node_client, piece_cache)
+    });
```

Comment:
Also these should be spawned onto a separate thread since cache writing will use disk I/O (there is `run_future_in_dedicated_thread` utility function in farmer already for this purpose).